library/core/src/num/int_sqrt.rs

```diff
@@ -0,0 +1,285 @@
+
+/// This array stores the [integer square roots][1] and remainders of each
+/// [`u8`](prim@u8) value. For example, `U8_ISQRT_WITH_REMAINDER[17]` will be
+/// `(4, 1)` because the integer square root of 17 is 4 and because 17 is 1
+/// higher than 4 squared.
+///
+/// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+/// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+/// Encyclopedia."
+const U8_ISQRT_WITH_REMAINDER: [(u8, u8); 256] = {
+    let mut result = [(0, 0); 256];
+
+    let mut n: usize = 0;
+    let mut isqrt_n: usize = 0;
+    while n < result.len() {
+        result[n] = (isqrt_n as u8, (n - isqrt_n.pow(2)) as u8);
+
+        n += 1;
+        if n == (isqrt_n + 1).pow(2) {
+            isqrt_n += 1;
+        }
+    }
+
+    result
+};
+
+/// Returns the [integer square root][1] of any [`u8`](prim@u8) input.
+///
+/// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+/// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+/// Encyclopedia."
+#[must_use = "this returns the result of the operation, \
+              without modifying the original"]
+#[inline(always)]
+pub const fn u8(n: u8) -> u8 {
+    U8_ISQRT_WITH_REMAINDER[n as usize].0
+}
+
+/// Returns the [integer square root][1] of any [`usize`](prim@usize) input.
+///
+/// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+/// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+/// Encyclopedia."
+#[must_use = "this returns the result of the operation, \
+              without modifying the original"]
+#[inline(always)]
+pub const fn usize(n: usize) -> usize {
+    #[cfg(target_pointer_width = "16")]
+    {
+        u16(n as u16) as usize
+    }
+
+    #[cfg(target_pointer_width = "32")]
+    {
+        u32(n as u32) as usize
+    }
+
+    #[cfg(target_pointer_width = "64")]
+    {
+        u64(n as u64) as usize
+    }
+}
+
+/// Generates an `i*` function that returns the [integer square root][1] of any
+/// **nonnegative** input of a specific signed integer type.
+///
+/// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+/// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+/// Encyclopedia."
+macro_rules! signed_fn {
+    ($SignedT:ident, $UnsignedT:ident) => {
+        /// Returns the [integer square root][1] of any **nonnegative**
+        #[doc = concat!("[`", stringify!($SignedT), "`](prim@", stringify!($SignedT), ")")]
+        /// input.
+        ///
+        /// # Safety
+        ///
+        /// This results in undefined behavior when the input is negative.
+        ///
+        /// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+        /// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+        /// Encyclopedia."
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        #[inline(always)]
+        pub const unsafe fn $SignedT(n: $SignedT) -> $SignedT {
```

Comment:
I've tried that and it makes the tests worse due to what I think is a bug. Currently, if the input is below zero, it casts to the unsigned version and gives the integer square root of that, which will fail the tests that require that `isqrt` panic and `checked_isqrt` return `None` rather than producing the integer square root of a large positive number.

The `debug_assert` causes the test executable itself to actually crash if the assert fails. Even a panic message included in the assert call will fail to be printed. This outcome would be very confusing.

I need to try to create a minimal reproduction of the issue.

---

I inserted `debug_assert!(n >= 0);` as the top line of the `unsafe` functions, then I changed the call to `crate::num::int_sqrt::$ActualT(self as $ActualT - 10) as $SelfT` to subtract 10 from the input, causing low numbers to become negative to fail the assert. I get this output after the tests are built:
```
running 20 tests

num::int_sqrt::i128::isqrt_extended ... F

num::int_sqrt::i16::isqrt_extended ... F

num::int_sqrt::i64::isqrt_extended ... F

num::int_sqrt::i32::isqrt_extended ... F
error: test failed, to rerun pass `-p core --test coretests`

Caused by:
  process didn't exit successfully: `/Users/chaitrex/dev/rust/build/aarch64-apple-darwin/stage0-std/aarch64-apple-darwin/release/deps/coretests-93e1b019e4f059ff int_sqrt --quiet -Z unstable-options --format json` (signal: 5, SIGTRAP: trace/breakpoint trap)
Build completed unsuccessfully in 0:00:28
```
There are usually three phases to the tests: the tests in `tests`, the tests in `benches`, and the doctests. This crashes during the first phase and the subsequent phases aren't run.

Reply:
I figured out the issue. It wasn't the `debug_assert`. It was actually a `crate::hint::assert_unchecked` condition becoming false.

Let's say n is `5_i32`. 10 is subtracted by my attempt to trigger the `debug_assert` making it `-5_i32`. That's cast to `u32::MAX - 4`, which has an integer square root that exceeds the optimization hint in `crate::hint::assert_unchecked(result <= MAX_RESULT);`.

I'll leave the `debug_assert` in there and try to add some comments about what a crashing test executable means.

Reply:
I did that, but in a slightly different manner. I used [`assert_unsafe_precondition!`](https://doc.rust-lang.org/nightly/core/macro.assert_unsafe_precondition.html) instead of `debug_assert!`.

`debug_assert!` looks at whether `core` is compiled in debug mode, and it's usually not, so it almost always is optimized away. On the other hand, `assert_unsafe_precondition!` looks at whether the user code is compiled in debug mode.

<details>
<summary>Altered code</summary>

```rust
        #[inline(always)]
        pub const unsafe fn $SignedT(n: $SignedT) -> $SignedT {
            assert_unsafe_precondition!(
                check_library_ub,
                concat!(
                    "`isqrt` internal error: Passing a negative number to `core::num::int_sqrt::",
                    stringify!($SignedT),
                    "` causes undefined behavior."
                ),
                (n: $SignedT = n) => n >= 0
            );

            $UnsignedT(n as $UnsignedT) as $SignedT
        }
```
</details>

Reply:
OK, changed to `debug_assert!`. Will push after some more work.